Lib: SMF: Add initial transitions to HSMs and add smf_set_handled()

glenn-andrews opened the pull request:
Brings SMF framework closer into alignment with accepted Hierarchical State Machine operation by:
1. Allowing 'programming by difference' by having some child states handle events and prevent propagation up to the parent run actions while others propagate events up to a common handler in a parent state.
2. Optionally allow initial transitions within a parent state to determine the most nested child state to transition to.
3. Adding a test case for `CONFIG_SMF_INITIAL_TRANSITION` and `smf_set_handled()`
4. Updating documentation for the new API (and fixing some references)

There was discussion in https://github.com/zephyrproject-rtos/zephyr/issues/55344 about not making the initial transition a Kconfig option, but I'm not sure of any way else of doing it without permanently adding a pointer to each `smf_state` entry, which is a problem for resource-constrained devices.

This does not fix https://github.com/zephyrproject-rtos/zephyr/issues/66341 but documentation has been updated to warn users of the issue.

glenn-andrews committed:
Lib: SMF: Add initial transition and smf_set_handled()

Brings SMF framework closer into alignment with accepted Hierarchical State
Machine operation by:
1. Allowing 'programming by difference' by having some child states handle
   events and prevent propagation up to the parent run actions while others
   propagate events up to a common handler in a parent state.
2. Optionally allow initial transitions within a parent state to determine
   the most nested child state to transition to.
3. Adding a test case for `CONFIG_SMF_INITIAL_TRANSITION` and
   `smf_set_handled()`
4. Updating documentation for the new API (and fixing some references)

There was discussion in https://github.com/zephyrproject-rtos/zephyr/issues/55344
about not making the initial transition a Kconfig option, but I'm not sure
of any way else of doing it without permanently adding a pointer to each
`smf_state` entry, which is a problem for resource-constrained devices.

This does not fix https://github.com/zephyrproject-rtos/zephyr/issues/66341
but documentation has been updated to warn users of the issue.

Signed-off-by: Glenn Andrews <glenn.andrews.42@gmail.com>

sambhurst commented on lib/smf/smf.c:
```diff
@@ -18,14 +18,12 @@ struct internal_ctx {
-	for (const struct smf_state *state = test_state;
-	     state != NULL;
-	     state = state->parent) {
+	for (const struct smf_state *state = test_state; state != NULL; state = state->parent) {
```
please run clang-format

glenn-andrews replied:
I did, and that was the result, plus messing up all the struct definitions.

Is there some magic I need to invoke on Windows other than `clang_format -i smf.c` as recommended by the contribution guidelines?

glenn-andrews replied:
On recommendation of the Discord, I've reverted all clang-format changes.

glenn-andrews replied:
> please run clang-format

Is the update satisfactory?

sambhurst replied:
Looks good to me

glenn-andrews commented on include/zephyr/smf.h:
```diff
@@ -136,6 +163,15 @@ void smf_set_state(struct smf_ctx *ctx, const struct smf_state *new_state);
+ *
+ * @param ctx  State machine context
+ */
+void smf_set_handled(struct smf_ctx *ctx);
```
I have a rather large sample that implements the PSiCC2 Section 2.3.15 demo app with all levels of transitions. It's about 11K of source, given the large number of states the demo app encompasses.

keith-zephyr replied:
That's probably too large.  I was thinking we had existing samples that use the SMF, but there are not.

However, there are some good unit tests for the SMF [here](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/lib/smf/src).

Please extend one of the existing tests or create a new one that enables the `CONFIG_SMF_INITIAL_TRANSITION` option.  The ensures that Zephyr's CI will validate that this new feature.

glenn-andrews replied:
Done. I've never used the unit tests before so feedback is appreciated.

glenn-andrews replied:
> That's probably too large. I was thinking we had existing samples that use the SMF, but there are not.

Would it be worth me submitting my demo app as a separate sample for SMF? I can probably simplify it a bit.

kartben replied:
> > That's probably too large. I was thinking we had existing samples that use the SMF, but there are not.
> 
> Would it be worth me submitting my demo app as a separate sample for SMF? I can probably simplify it a bit.

Yes, please!

glenn-andrews commented on doc/services/smf/index.rst:
```diff
@@ -39,6 +39,10 @@ By default, a state can have no ancestor states, resulting in a flat state
 
+By default, the hierarchical state machine does not support initial transitions
+to child states on entering a superstate. To enable them the
+:kconfig:option:`CONFIG_SMF_INITIAL_TRANSITION` option must be enabled.
```
I tried my hand at a graphviz statechart. Couldn't wok out how to get an initial transition from the outside to STATE_S0. Let me know if it's OK as is.

glenn-andrews replied:
The graphviz appears to be having trouble. Can you see the issue? I've never used it before.

kartben replied:
> The graphviz appears to be having trouble. Can you see the issue? I've never used it before.

you fixed it, yes? FWIW I couldn't find a solution for drawing the transition symbolizing that S0 is the entry state :-/

keith-zephyr replied:
This gets close to my understanding the transitions for the state machine defined by your test:
```
.. graphviz::
   :caption: demo_states with initial transition to S2

   digraph smf_hierarchical_initial {
      compound=true;
      node [style = rounded];
      smf_set_initial [shape=plaintext];
      ab_init_state [shape = point];
      STATE_A [shape = box];
      STATE_B [shape = box];
      STATE_C [shape = box];
      STATE_D [shape = box];

      subgraph cluster_ab {
         label = "PARENT_AB";
         style = rounded;
         ab_init_state -> STATE_A;
         STATE_A -> STATE_B;
      }

      subgraph cluster_c {
         label = "PARENT_C";
         style = rounded;
	 STATE_C -> STATE_C
      }

      smf_set_initial -> STATE_A [lhead=cluster_ab]
      STATE_B -> STATE_C
      STATE_C -> STATE_D
   }
```

glenn-andrews commented on doc/services/smf/index.rst:
```diff
@@ -71,6 +75,21 @@ And this example creates three hierarchical states::
+      [S2] = SMF_CREATE_STATE(s2_entry, s2_run, s2_exit, demo_states[S0], NULL)
+   };
+
+
```
See new note.

glenn-andrews commented on doc/services/smf/index.rst:
```diff
@@ -317,7 +343,11 @@ When designing hierarchical state machines, the following should be considered:
-   without transitioning to another state, ie. calling smf_set_state.
+   without transitioning to another state, ie. calling smf_set_state, or if
+   ``smf_set_handled`` is called within the child_run function.
+ - When a parent state intitiates a transition to self, the parents's exit
+   action is not called, e.g. instead of child_exit, parent_exit, parent_entry
+   it performs child_exit, parent_entry
```
Do we want to be testing to confirm a bug exists? I don't want to see this codified as "the way things work" and add additional effort to fixing it by having to fix the test.

keith-zephyr replied:
The documentation here defines an API contract:  A parent state initiating a transition to self does not call the parent_exit.  I didn't consider this a bug.
So I wondering if this condition could explicitly be tested.

glenn-andrews replied:
It's considered a bug here: https://github.com/zephyrproject-rtos/zephyr/issues/66341

glenn-andrews commented on tests/lib/smf/src/test_lib_initial_transitions_smf.c:
```diff
@@ -0,0 +1,501 @@
+	test_obj.transition_bits = 0;
+	test_obj.first_time = 1;
+	test_obj.terminate = PARENT_RUN;
+	smf_set_initial((struct smf_ctx *)&test_obj, &test_states[STATE_A]);
```
Oops. Fixed.

kartben commented on doc/services/smf/index.rst:
```diff
@@ -79,6 +117,13 @@ To transition from one state to another, the ``smf_set_state`` function is
 used and it has the following prototype:
 ``void smf_set_state(smf_ctx *ctx, smf_state *state)``
 
+**NOTE:** If :kconfig:option:`CONFIG_SMF_INITIAL_TRANSITION` is not set,
```
There is a `.. note::` Sphinx directive that would probably be appropriate here and in other places with `**NOTE**`s

keith-zephyr commented on doc/services/smf/index.rst:
```diff
@@ -71,6 +78,37 @@ And this example creates three hierarchical states::
    };
 
 
+This example creates three hierarchical states with an initial transition
+from parent state S0 to child state S2::
```
This section should be added following the Hierarchical State Machine Example.  It should also include the code showing state transitions.

As is, I don't think this diagram is helpful to the user as it's demonstrating too trivial a case.

But I'm okay with you dropping this example/diagram from this PR and creating a follow-up PR to make the improvements here.

keith-zephyr replied:
I think a good goal here would be to have a diagram that matches the state transitions from the test.

glenn-andrews replied:
Moved diagram to end of document.

glenn-andrews commented on doc/services/smf/index.rst:
```diff
@@ -91,6 +110,13 @@ some application dependent way. An application should cease calling
+Preventing Parent Run Actions
+=============================
+
+Calling ``smf_set_handled`` prevents calling the run action of parent states.
+It is not required to call ``smf_set_handled`` if the state calls
+``smf_set_state``.
```
Do you want me to refactor the entire document, or just the parts I've touched?

glenn-andrews commented:
Anything else I should do?

kartben reviewed (changes requested):
@glenn-andrews thanks much for your work on this! I've pushed a couple commits to your branch adding what was missing for the functions to properly render as links.

The one thing that I think needs to be changed is that you have *all* your changes into just one commit (which description now seems outdated/incomplete anyway?), which is not ideal. It would be great if you could split in several commits so that the additions of new Kconfigs and APIs are not mixed with what are "simple" documentation enhancements.

Thanks!

glenn-andrews commented:
> It would be great if you could split in several commits so that the additions of new Kconfigs and APIs are not mixed with what are "simple" documentation enhancements.

I'm not entirely sure how such a split would work. We've got about 8 instances where I replaced double-backticks with the appropriate sphinx directives, all the rest is directly relevant to the API being added. I could revert the changes to the double-backticks, then artificially add them again, but that seems silly.

Can you give me a clearer idea of what is required? I feel I'm missing the intent of this request.

glenn-andrews committed:
Lib: SMF: Fix `smf_set_terminate` function typo.

Fixed where `smf_set_terminate` was written `smf_terminate`

Signed-off-by: Glenn Andrews <glenn.andrews.42@gmail.com>

glenn-andrews commented:
> I've pushed a couple commits to your branch adding what was missing for the functions to properly render as links.

I'm terribly sorry, but I think I managed to nuke your changes by not pulling them before I amended my commit message. I fixed the `smf_terminate` one, but can't remember where the doxygen one goes.

glenn-andrews committed:
Lib: SMF: Add Doxygen commands.

Added Doxygen markup.

Signed-off-by: Glenn Andrews <glenn.andrews.42@gmail.com>

glenn-andrews commented:
How's this looking now? Any more action items for me?

glenn-andrews committed:
Lib: SMF: Update index.rst

Incorporated changes requested by @keith-zephyr:
1. Grammar fix
2. Make explicit that transition to self in super-states is not supported

Signed-off-by: Glenn Andrews <glenn.andrews.42@gmail.com>

glenn-andrews commented:
https://github.com/zephyrproject-rtos/zephyr/pull/66753#discussion_r1510029400

Updated. Does the new wording seem okay?

kartben reviewed (approved):
Thanks a lot for all the hard work, @glenn-andrews!

glenn-andrews commented:
Thank you all so much for reviewing this!

bryceschober commented on doc/services/smf/index.rst:
```diff
@@ -466,3 +504,48 @@ Code::
+The statechart for this test is below.
+
+.. graphviz::
+   :caption: Test state machine for initial trnasitions and ``smf_set_handled``
```
Typo "trnasitions"
